core/src/main/scala/io/snappydata/SnappyThinConnectorTableStatsProvider.scala

```diff
@@ -108,23 +111,29 @@ object SnappyThinConnectorTableStatsProvider extends TableStatsProviderService {
   }
 
   override def getStatsFromAllServers(sc: Option[SparkContext] = None): (Seq[SnappyRegionStats],
-      Seq[SnappyIndexStats], Seq[SnappyExternalTableStats]) = {
+      Seq[SnappyIndexStats], Seq[SnappyExternalTableStats]) = synchronized {
     try {
-      executeStatsStmt(sc)
+      val resultSet = executeStatsStmt(sc)
+      val regionStats = new ArrayBuffer[SnappyRegionStats]
+      while (resultSet.next()) {
+        val tableName = resultSet.getString(1)
+        val isColumnTable = resultSet.getBoolean(2)
+        val isReplicatedTable = resultSet.getBoolean(3)
+        val rowCount = resultSet.getLong(4)
+        val sizeInMemory = resultSet.getLong(5)
+        val totalSize = resultSet.getLong(6)
+        regionStats += new SnappyRegionStats(tableName, totalSize, sizeInMemory, rowCount,
+          isColumnTable, isReplicatedTable)
+      }
+      (regionStats, Nil, Nil)
     } catch {
-      case e: Exception =>
+      case NonFatal(e) =>
         logWarning("Warning: unable to retrieve table stats " +
             "from SnappyData cluster due to " + e.toString)
         logDebug("Exception stack trace: ", e)
         closeConnection()
```

Comment:
I am not sure closing connection here is a good idea. It might be some temporary unavailability.

Reply:
Probably should re-initialize connection for safety.

Reply:
Yes, connection is reinitialized by "executeStatsStmt" if it has been closed.